File: windows-app/Plantoir.Tests/ClassFolderContractTests.cs
using System.Text.Json.Nodes;
using Plantoir.Core.Models;

namespace Plantoir.Tests;

/// <summary>
/// The class-folder rule, run against the SHARED contract.
///
/// <para>The cases are deserialised from <c>contracts/class-planning.json</c> →
/// <c>classFolder</c>, never retyped — the same data the macOS suite and
/// <c>scripts/test_class_folder.py</c> run against their own implementations.
/// That is the point: this rule used to exist four times and disagree, and this
/// app's copy was the one that tested the whole directory string, so a teacher
/// whose working folder was <c>C:\Users\x\Classroom\</c> made every page in
/// every course a class page.</para>
/// </summary>
public class ClassFolderContractTests
{
    [Fact]
    public void ClassFolderNaming_MatchesContract()
    {
        var doc = ContractLoader.LoadJson("class-planning.json");
        var cases = doc["classFolder"]!["naming"]!["cases"]!.AsArray();
        Assert.True(cases.Count >= 5, "the contract lost naming cases");

        foreach (var c in cases)
        {
            if (c is null) continue;
            var folders = c["perSectionFolders"]!.AsArray().Select(x => x!.ToString()).ToList();
            string expected = c["expect"]!.ToString();
            string name = c["name"]?.ToString() ?? "unnamed";
            // A case WITHOUT `classFolder` is a course that never recorded
            // one, so it is read with a default rather than treated as a new
            // shape — the same trap as pageNaming's `term`.
            string? recorded = c["classFolder"]?.ToString();

            Assert.True(ClassFolderRule.Name(recorded, folders) == expected,
                $"{name}: expected '{expected}', got '{ClassFolderRule.Name(recorded, folders)}'");
        }
    }

    [Fact]
    public void ClassFolderMembership_MatchesContract()
    {
        var doc = ContractLoader.LoadJson("class-planning.json");
        var cases = doc["classFolder"]!["membership"]!["cases"]!.AsArray();
        Assert.True(cases.Count >= 4, "the contract lost membership cases");

        foreach (var c in cases)
        {
            if (c is null) continue;
            var folders = c["perSectionFolders"]!.AsArray().Select(x => x!.ToString()).ToList();
            var expected = c["expect"]!.AsArray().Select(x => x!.ToString()).ToList();
            string name = c["name"]?.ToString() ?? "unnamed";
            string? recorded = c["classFolder"]?.ToString();

            Assert.True(ClassFolderRule.Names(recorded, folders).SequenceEqual(expected),
                $"{name}: expected [{string.Join(", ", expected)}], got " +
                $"[{string.Join(", ", ClassFolderRule.Names(recorded, folders))}]");
        }
    }

    [Fact]
    public void IsClassPage_MatchesContract()
    {
        var doc = ContractLoader.LoadJson("class-planning.json");
        var cases = doc["classFolder"]!["isClassPage"]!["cases"]!.AsArray();
        Assert.True(cases.Count >= 12, "the contract lost isClassPage cases");

        foreach (var c in cases)
        {
            if (c is null) continue;
            string path = c["path"]!.ToString();
            var folders = c["classFolders"]!.AsArray().Select(x => x!.ToString()).ToList();
            bool expected = c["expect"]!.GetValue<bool>();
            string name = c["name"]?.ToString() ?? "unnamed";

            Assert.True(ClassFolderRule.IsClassPage(path, folders) == expected,
                $"{name}: {path} against [{string.Join(", ", folders)}] should be {expected}");
        }
    }

    /// <summary>
    /// Defence in depth, and labelled as such: under segment EQUALITY these
    /// could not match anyway. The test exists so a future change to prefix or
    /// substring matching fails here rather than quietly reclassifying content
    /// that ships in the example payloads.
    /// </summary>
    [Fact]
    public void APageIsNeverALessonBecauseOfItsName()
    {
        var folders = new[] { "All Classes" };
        foreach (var page in new[]
                 {
                     "Setup/How This Class Works.md",
                     "Setup/Our Classroom Norms.md",
                     "Curriculum/B3. Connections Beyond the Classroom.md",
                     "All Classes.md",
                 })
        {
            Assert.False(ClassFolderRule.IsClassPage(page, folders), page);
        }
    }

    /// <summary>
    /// The regression that matters most on THIS platform: Plan() used to test
    /// the whole ABSOLUTE directory string, so a working folder called
    /// C:\Users\x\Classroom made every page in every course a class page.
    /// The rule is a pure segment matcher, so what protects it is entirely
    /// what Plan() PASSES it — <c>AssistWorkspace.PathWithinSection(...)</c>,
    /// the page's path relative to its section, since 2026-09-19; it passed
    /// <c>Relative(pagePath)</c> before that, which closed this bug and left
    /// <c>courses</c>, the course code and <c>sectionN</c> still above it.
    /// Asserted here as the SHAPE the rule expects; that Plan() really hands
    /// it that shape is <c>ClassFolderMembershipTests</c>, which builds the
    /// folder tree and asks through a real workspace.
    /// </summary>
    [Fact]
    public void OnlyARelativePathIsEverAskedAbout()
    {
        var folders = new[] { "All Classes" };
        Assert.False(ClassFolderRule.IsClassPage(@"Concepts\Loops.md", folders));
        Assert.True(ClassFolderRule.IsClassPage(@"All Classes\Unit 1, Day 1.md", folders));
        // A page whose ancestors mention classes, once the path is cut back to
        // the section, is not a lesson — which is the whole point of passing a
        // narrowed path rather than an absolute one.
        Assert.False(ClassFolderRule.IsClassPage(@"Concepts\Recursion.md", folders));
    }

    [Fact]
    public void ANullFolderEntryDoesNotThrow()
    {
        // These lists come from JSON, including the contract's own case data.
        Assert.Equal("All Classes", ClassFolderRule.Name(new string?[] { null }!));
        Assert.False(ClassFolderRule.IsClassPage("All Classes/x.md", new string?[] { null }!));
    }
}

File: mac-app/Tests/QuartzTeachersTests/ClassFolderContractTests.swift
import XCTest
@testable import QuartzTeachers

/// The class-folder rule, run against the SHARED contract.
///
/// The cases are deserialised from `contracts/class-planning.json` →
/// `classFolder`, never retyped — the same data `scripts/test_class_folder.py`
/// and the Windows suite run against their own implementations. That is the
/// whole point: this rule used to exist four times and disagree, and the
/// build's disagreement silently changed the Curriculum Coverage map from
/// "pages the course teaches" to "every published page".
@MainActor
final class ClassFolderContractTests: XCTestCase {

    // MARK: - Stored properties

    private var contract: [String: Any] = [:]

    // MARK: - Functions

    override func setUpWithError() throws {
        // Located the way ClassPlanningContractTests locates the same file:
        // up from #filePath to the repository root.
        let url: URL = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent().deletingLastPathComponent()
            .deletingLastPathComponent().deletingLastPathComponent()
            .appendingPathComponent("contracts/class-planning.json")
        let all: [String: Any] = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: try Data(contentsOf: url)) as? [String: Any]
        )
        contract = try XCTUnwrap(all["classFolder"] as? [String: Any],
                                 "contracts/class-planning.json has no classFolder section")
    }

    func testNamingCasesFromTheContract() throws {
        let naming: [String: Any] = try XCTUnwrap(contract["naming"] as? [String: Any])
        let cases: [[String: Any]] = try XCTUnwrap(naming["cases"] as? [[String: Any]])
        XCTAssertGreaterThanOrEqual(cases.count, 10, "the contract lost naming cases")
        for testCase in cases {
            let name: String = (testCase["name"] as? String) ?? "unnamed"
            let folders: [String] = (testCase["perSectionFolders"] as? [String]) ?? []
            let expected: String = try XCTUnwrap(testCase["expect"] as? String)
            XCTAssertEqual(
                // A case with no `classFolder` is a course that has never
                // recorded one — every course made before the key existed —
                // and must still get the old guess.
                ClassFolder.name(
                    inPerSectionFolders: folders, configured: testCase["classFolder"] as? String
                ), expected,
                "\(name): \((testCase["why"] as? String) ?? "")"
            )
        }
    }

    func testIsClassPageCasesFromTheContract() throws {
        let rule: [String: Any] = try XCTUnwrap(contract["isClassPage"] as? [String: Any])
        let cases: [[String: Any]] = try XCTUnwrap(rule["cases"] as? [[String: Any]])
        XCTAssertGreaterThanOrEqual(cases.count, 12, "the contract lost isClassPage cases")
        for testCase in cases {
            let name: String = (testCase["name"] as? String) ?? "unnamed"
            let path: String = try XCTUnwrap(testCase["path"] as? String)
            let folders: [String] = try XCTUnwrap(testCase["classFolders"] as? [String])
            let expected: Bool = try XCTUnwrap(testCase["expect"] as? Bool)
            XCTAssertEqual(
                ClassFolder.isClassPage(relativePath: path, classFolders: folders), expected,
                "\(name): \((testCase["why"] as? String) ?? "")"
            )
        }
    }

    func testMembershipCasesFromTheContract() throws {
        let rule: [String: Any] = try XCTUnwrap(contract["membership"] as? [String: Any])
        let cases: [[String: Any]] = try XCTUnwrap(rule["cases"] as? [[String: Any]])
        XCTAssertGreaterThanOrEqual(cases.count, 6, "the contract lost membership cases")
        for testCase in cases {
            let name: String = (testCase["name"] as? String) ?? "unnamed"
            let folders: [String] = (testCase["perSectionFolders"] as? [String]) ?? []
            let expected: [String] = try XCTUnwrap(testCase["expect"] as? [String])
            XCTAssertEqual(
                ClassFolder.names(
                    inPerSectionFolders: folders, configured: testCase["classFolder"] as? String
                ), expected,
                "\(name): \((testCase["why"] as? String) ?? "")"
            )
        }
    }

    /// Defence in depth, and labelled as such: under segment EQUALITY these
    /// could not match anyway. The test exists so a future change to prefix or
    /// substring matching fails here rather than quietly reclassifying content
    /// that ships in the example payloads.
    func testAPageIsNeverALessonBecauseOfItsName() {
        let pages: [String] = [
            "Setup/How This Class Works.md",
            "Setup/Our Classroom Norms.md",
            "Curriculum/B3. Connections Beyond the Classroom.md",
            "All Classes.md",
        ]
        for page in pages {
            XCTAssertFalse(
                ClassFolder.isClassPage(relativePath: page, classFolders: ["All Classes"]),
                "\(page) is not a lesson — a page is never one because of its NAME"
            )
        }
    }

    /// The bug an adversarial review found still live on THIS platform.
    ///
    /// `AssistSectionPage.relativePath` is relative to the working folder — and
    /// is the FULL ABSOLUTE PATH when `workspaceURL` is nil, which
    /// `SectionIndexPointer.repointIndex` passes. Asking the class-page
    /// question of that string meant a teacher whose working folder was named
    /// for classes made every page in every course a lesson.
    ///
    /// The rule itself is a pure segment matcher and cannot know what sits
    /// above the content root, so what protects it is the page carrying
    /// `pathWithinSection`. That is what this exercises — a working folder
    /// called "All Classes", which under the old code made every page a class
    /// page.
    func testAWorkingFolderNamedForClassesCannotMakeEveryPageALesson() throws {
        let root: URL = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("All Classes")
            .appendingPathComponent("class-folder-\(UUID().uuidString)")
        let courseURL: URL = root.appendingPathComponent("courses/ICS3U")
        let sectionURL: URL = courseURL.appendingPathComponent("section1")
        try FileManager.default.createDirectory(
            at: sectionURL.appendingPathComponent("Concepts"), withIntermediateDirectories: true
        )
        defer { try? FileManager.default.removeItem(at: root) }

        let configuration: [String: Any] = [
            "course_code": "ICS3U", "course_name": "Introduction to Computer Science",
            "section_numbers": [1], "num_sections": 1,
            "per_section_folders": ["All Classes"], "per_section_files": [],
        ]
        let configURL: URL = courseURL.appendingPathComponent("course_config.json")
        try JSONSerialization.data(withJSONObject: configuration, options: [.prettyPrinted])
            .write(to: configURL)
        let course: Course = Course(
            code: "ICS3U", directoryURL: courseURL,
            configuration: try CourseConfiguration(contentsOf: configURL)
        )

        let lesson: URL = sectionURL.appendingPathComponent("Concepts/Loops.md")
        let within: String = AssistSectionGraph.pathWithinSection(
            of: lesson, forSection: 1, in: course
        )
        XCTAssertEqual(within, "Concepts/Loops.md",
                       "everything above the section folder must be stripped")
        XCTAssertFalse(
            ClassFolder.isClassPage(relativePath: within, classFolders: ClassFolder.names(for: course)),
            "a working folder named 'All Classes' must not make Loops.md a lesson"
        )

        let realClass: URL = sectionURL.appendingPathComponent("All Classes/Unit 1, Day 1.md")
        XCTAssertTrue(ClassFolder.isClassPage(
            relativePath: AssistSectionGraph.pathWithinSection(of: realClass, forSection: 1, in: course),
            classFolders: ClassFolder.names(for: course)
        ))
    }

    func testAWindowsPathSeparatorIsUnderstood() {
        XCTAssertTrue(ClassFolder.isClassPage(
            relativePath: "All Classes\\Unit 1, Day 1.md", classFolders: ["All Classes"]))
    }
}
